query-node/mappings/src/bounty.ts

```diff
@@ -0,0 +1,485 @@
+  if (!bounty) {
+    throw new Error(`Bounty not found by id: ${bountyId}`)
+  }
+  return bounty
+}
+
+async function getContribution(
+  store: DatabaseManager,
+  bountyId: BountyId,
+  contributor?: string
+): Promise<BountyContribution> {
+  const contribution = await store.get(BountyContribution, { where: { bountyId, contributor } })
+  if (!contribution) {
+    const actorType = typeof contributor === 'undefined' ? 'council' : `member id ${contributor}`
+    throw new Error(`Bounty contribution not found by contributor: ${actorType}`)
+  }
+  return contribution
+}
+
+async function getEntry(store: DatabaseManager, entryId: EntryId): Promise<BountyEntry> {
+  const entry = await store.get(BountyEntry, { where: { id: entryId } })
+  if (!entry) {
+    throw new Error(`Entry not found by id: ${entryId}`)
+  }
+  return entry
+}
+
+async function updateBounty(
+  store: DatabaseManager,
+  event: SubstrateEvent,
+  bountyId: BountyId,
+  changes: (bounty: Bounty) => Partial<Bounty>
+) {
+  const bounty = await getBounty(store, bountyId)
+  bounty.updatedAt = new Date(event.blockTimestamp)
+  Object.assign(bounty, changes(bounty))
+
+  await store.save<BountyEntry>(bounty)
+
+  return bounty
+}
+
+async function updateEntry(
+  store: DatabaseManager,
+  event: SubstrateEvent,
+  entryId: EntryId,
+  changes: (bounty: BountyEntry) => Partial<BountyEntry>
+) {
+  const entry = await getEntry(store, entryId)
+  entry.updatedAt = new Date(event.blockTimestamp)
+  Object.assign(entry, changes(entry))
+
+  await store.save<BountyEntry>(entry)
+
+  return entry
+}
+
+function bountyActorToMembership(actor: BountyActor): Membership | undefined {
+  if (actor.isMember) {
+    return new Membership({ id: String(actor.asMember) })
+  }
+}
+
+function fundingPeriodEnd(bounty: Bounty): number {
+  return (
+    bounty.maxFundingReachedEvent?.inBlock ??
+    bounty.createdInEvent.inBlock + (bounty.fundingType as BountyFundingLimited).fundingPeriod
+  )
+}
+
+function whenDef<T, R>(value: T | null | undefined, fn: (value: T) => R): R | undefined {
+  if (value !== null && typeof value !== 'undefined') return fn(value)
+}
+
+/**
+ * Schedule Periods changes
+ */
+
+export function bountyScheduleFundingEnd(store: DatabaseManager, bounty: Bounty): void {
+  const { fundingType } = bounty
+  if (bounty.stage !== BountyStage.Funding || !('fundingPeriod' in fundingType)) return
+
+  const fundingPeriodEnd = bounty.createdInEvent.inBlock + fundingType.fundingPeriod
+  scheduleAtBlock(fundingPeriodEnd, () => {
+    if (bounty.stage === BountyStage.Funding) {
+      const isFunded = bounty.totalFunding >= fundingType.minFundingAmount
+      endFundingPeriod(store, bounty, isFunded)
+    }
+  })
+}
+
+export function bountyScheduleWorkSubmissionEnd(store: DatabaseManager, bounty: Bounty): void {
+  if (bounty.stage !== BountyStage.WorkSubmission) return
+
+  const workingPeriodEnd = fundingPeriodEnd(bounty) + bounty.workPeriod
```

Comment:
This is incorrectly calculated, because `bounty.maxFundingReachedEvent.inBlock` is not yet available here when the function is called from `bounty_BountyMaxFundingReached`.

I added some console logs to ilustrate this (the bounty funding type was `Perpetual` and the funding target was reached):
```
[36mprocessor                |[0m Calculating funding period end {
[36mprocessor                |[0m   maxFundingReachedEventBlock: undefined,
[36mprocessor                |[0m   bountyCreatedInEventBlock: 123,
[36mprocessor                |[0m   bountyFundingPeriod: undefined
[36mprocessor                |[0m }
[36mprocessor                |[0m Scheduling work submission end at NaN
```